images/router/haproxy/reload-haproxy

```diff
@@ -15,7 +16,40 @@ if [ -f $pid_file ]; then
 fi
 
 if [ -n "$old_pid" ]; then
+  if $(set | grep DROP_SYN_DURING_RESTART= > /dev/null) && [ -n "$DROP_SYN_DURING_RESTART" ]; then
+    # We install the syn eater so that connections that come in during the restart don't
+    # go onto the wrong socket, which is then closed.  The sleep is needed to let the socket
+    # drain before the new haproxy starts and binds to the same port.
+    ports=$(grep -E '^\s*bind\s+:[[:digit:]]+\w' "$config_file" | cut -f2 -d: | paste -d, -s)
+    if [ -n "$ports" ]; then
+      # If this doesn't insert, we don't care, we still want to reload
+      /usr/sbin/iptables -I INPUT -p tcp -m multiport --dports $ports --syn -j DROP \
+			 -m comment --comment "Eat SYNs while reloading haproxy" || :
+      installed_iptables=1
+      sleep 0.1
```

Comment:
I couldn't think of one.  Basically we need to drop new connections long enough for the running haproxy to accept all of the ones on the incoming socket.  But I don't think we can see whether they have been accepted.  I don't think the distinction between ESTABLISHED, SYN_SENT, and SYN_RECV is at the right level.

I agree that an arbitrary timeout is scary.  I tested and hit errors with a sleep as low as 0.01 (with other deliberate load on the system to make it less responsive).  So I figured a 10x buffer would cover most cases.  At worst, we may still see connection resets, but that's what happens today.  The ideal fix is for haproxy to have a reliable reload mechanism, or for the kernel to do the right thing when a process dies with unaccepted connections when there's another program bound to the same socket (with SO_REUSE set).  The former is not looking likely, but people in Rashid's group are looking at the latter.  It apparently has been a long-standing request, so appears to be getting some attention finally.